docs/Timelines.md

```diff
@@ -0,0 +1,149 @@
+it contributes to the overall picture.
+
+# Fake timeline: your first timeline
+
+Before timelines, Nakadi would connect to a single Kafka cluster,
+which used to be specified in the application yaml file. This was a
+static configuration loaded during boot time. Once timelines were
+introduced, in order to ease the migration process and remove special
+cases from the implementation, a fake timeline was introduced.
+
+Fake timelines exist only in memory. They are Timelines that are not
+stored in the database. Fake timelines can be seen as a fallback
+strategy for when there is no actual timeline for a given event type.
+
+Also, fake timelines use the default storage defined in the
+application configuration.
+
+# Timeline creation
+
+Timeline creation is coordinated through a series of locks and
+barriers using Zookeeper. Following we depict an example of what the
+ZK datastructure looks like at each step.
+
+## Initial state
+
+Every time a Nakadi application is launched, it tries to create the
+following ZK structure:
+
+```
+- timelines
+  + - lock                    lock for timeline versions synchronization
+  + - version: {version}      monotonically incremented long value (version of timelines configuration)
+  + - locked_et
+  + - nodes                   nakadi nodes
+    + - {node1}: {version}    Each nakadi node exposes the version used on this node
+    + - {node2}: {version}
+```
+
+In order to not override the initial structure, due to concurrency,
+each instance needs to take the lock `/nakadi/timelines/lock` before
+executing.
+
+## Start timeline creation for et_1
+
+When a new timeline creation is initiated, the first step is to
+acquire a lock to update timelines for et_1 by creating an ephemeral
+node at `/timelines/locked_et/et_1`.
+
+```
+- timelines
+  + - lock
+  + - version: 0
+  + - locked_et:
+    + - et_1
+  + - nodes
+    + - node1: 0
+    + - node2: 0
+```
+
+## Notify all Nakadi nodes about change: the version barrier
+
+Next, the instance coordinating the timeline creation bumps the
+version node, which all Nakadi instances are listening to changes, so
+they are notified when something changes.
+
+```
+- timelines
+  + - lock
+  + - version: 1       # this is incremented by 1
+  + - locked_et:
+    + - et_1
+  + - nodes
+    + - node1: 0
+    + - node2: 0
+```
+
+## Wait for all nodes to react to the new version
+
+Each Nakadi instance watches the value of the
+`/nakadi/timelines/version/` node. When it changes, each instance
+checks all locked event types and reacts accordingly, by either
+releasing or blocking publishers locally.
+
+Once each instance has updated its local list of locked event types,
+it bumps its own version, to let the timeline creator initiator know
+that it can proceed.
+
+```
+- timelines
+  + - lock
+  + - version: 1
+  + - locked_et:
+    + - et_1
+  + - nodes
+    + - node1: 1       # each instance updates its own version
+    + - node2: 1
+```
+
+## Proceed with timeline creation
```

Comment:
I think that it makes sense to add section regarding fake timelines and switch to very first one?
The main difference is that for very first not fake timeline topic is not created. Instead, old one is used. For example, suppose that we have et et1. , that is linked to topic topic1
```
et1  -> [FakeTimeline(topic1)]
# then, when we try to create first timeline, situation will actually change to 
et1 -> [RealTimeline(topic1)]
# Old storage is is used for that request
# And when new timeline for topic created again, than we finally have situation with 2 timelines
et1-> [RealTimeline(topic1), RealTimeline(topic2)]
```
The purpose of replacing fake to first real timeline is to switch cursor format. 
Even more - very first real timeline can be rolled back (therefore cursor format will switch to old format)